### ISSUE TITLE
Notification for none-responding devices

### ISSUE THREAD

**Issue by sbouchex** [domoticz/domoticz#829]
Idea : I'd like to be notified when a device is not responding (because it's dead, the line is cut...)

Implementation:
- Option 1 : A global parameter + A checkbox per device to disable this feature
- Option 2 : A notification "Deadman" to per with a customer message...

**Comment by gizmocuz** [comment]
This is only for zwave related hardware i suppose, because you call it 'dead'
How often do you experience this ? Is it not a mesh problem ?

**Comment by sbouchex** [comment]
The word "dead" is not related to zwave but rather the monitoring world and rails world (see https://en.wikipedia.org/wiki/Dead_man%27s_switch)

A tool like Kapacitor is using it (see https://github.com/influxdata/kapacitor/pull/149) and it can be use the detect faulty or none-responding sensors

zwave, mqtt, HTTP poller and any hardware doing queries can use this mechanism.

**Comment by gizmocuz** [comment]
Let me rephrase my comment:
- from what hardware you are receiving dead nodes ?

Do note, that domoticz has a 'sensor timeout' notification if a sensor is not received for xx seconds/hours

I am also not a big fan of the word 'dead', but that might have todo with a recent personal happening

**Comment by sbouchex** [comment]
Sorry for the strong wording, it's more a technical word used elsewhere and I understand it can heard.

In fact, any zwave, mqtt, HTTP poller or any hardware doing queries can have this kind of 'none-responding' issues

the "sensor timeout" can be a solution but it's just "ERROR" logging, however I would prefer an additional notification. Why not adding this kind of thing as well... ?

**Comment by gizmocuz** [comment]
If you have a stale sensor, this means, there is no data coming in to domoticz (consider it dead, cable is out , no power etc)
You can enable a notification for this

If a hardware class becomes unresponsive, we already have a system for this, and it will auto restart (threading code thing)

You can even (at the hardware setup level), specify a data timeout, if there is no data received for an amount of time, the hardware will be restarted (good if for example mqtt/http crashes)

So we have 3 options...

**Comment by RicardP** [comment]
I would say an option for user to have notifications sent from the already available Error Log would be great function? I am also missing to get notified if something goes wrong in Domoticz, I believe "user control" should be the main thing when building a home automation system. First work on system and devices up-time "rock solid stability", then features. 

I started with 433 devices but understood very soon it is nothing to use in a serious project so I went to ZWave. And yes of course it is better, but not as reliable as it is meant to be, I suppose. The multi manufacture licensing is great as it gives many products and competition, however there seems to be some problems with compatibility between different manufactures despite the standard. So, because of the weak spots in any system I would be happy to get notified if something unexpected happen or did not happen. I think a way of using the Error Log, with user filtering settings, would be the best base for this notification function.

I am impressed by all work, but I don't really trust the system as it is performing till now. Having my donation will need improvements in "system control and robustness". Would love that! :)

**Comment by sbouchex** [comment]
@RicardP agree with you, I would be interesting to have an option to sent "ERROR" log to notifications. We are not watching the log everyday but we can enable notifications to various systems like email, slack.io, SMS when something goes wrong.

@gizmocuz I have tested the options you are proposing but when a device is not responding and the option is activate, an ERROR log is added to the log but no notification is emitted.

**Comment by gizmocuz** [comment]
I think this function should handle this (but it's not for lighting sensors as they could not be touched for days)

//Executed every hour
void CSQLHelper::CheckDeviceTimeout()

But it could also be that it is not enabled in domoticz (the 'SensorTimeoutNotification' preference is not set anywhere)

@RichardP, I agree that the ZWave setup is rather complex and you could also easily get problems when for example your mesh network is not good.
This does not mean Domoticz is an unreliable system, might be for 'you'

Depending on your hardware platform, if you use Linux (on a raspberry for example), you can log to syslog, and if you receive errors, you could forward them via email.
But that could be rather complex to setup.

If you expect errors, i would watch the error logs (via the gui) everyday now

Forwarding them to email seems like a good idea, what happens when there is an error every minute (a hardware sensor is unable to be polled, or a connection could not be made), maybe we should make an internal queue, collect messages , and every 5 minutes check if there are messages to be send.
I can create this, but i will support only email for the moment as the messages could become rather large
But this is outside this topic

**Comment by gizmocuz** [comment]
Notification for Errors #832

**Comment by RicardP** [comment]
@gizmocuz, big thanks for this positive news and taking your time to work on this.

I dont mean ZWave is bad and I don't have any issues at the moment (went from Raspberryt to Windows as its more familiar to me), however especially a wireless system have potential issues that a wired wont, blocking walls, other 868 equipment interference, nodes radios/transmitter goes defective, bad internal antenna designs, etc. KNX by wire is a better system in stability but it is really only an option in new buildings and it is not flexible due to the same reason. The main problem with ZWave as I read on OWZ is that you cant (at least not with OZW) cant tell the controller to check/poll non responsive nodes (so called dead nodes). I also experienced the I had a power loss to PAN04-01 switch but after that the node never (several days) came back as active, not even if manually switching this node on/off/on/off. So I thought it was damaged due to some spike from the power grid... However then I had to reboot the raspberry at some point and then the "dead node" was back! So this behavior was not a node issue but the controller, much worse! I have to say I have been struggling with the Razberry controller so I cant blame Sigma Design for this, but at the same time OZW say they can do anything about "dead nodes" from software side...

Any system shall have error notification function due to the fact electronics stops working over time and again, I am happy to hear this is important function are looked into and the first version by e-mail is just fine. later on, if needed, it can be refined to be sent by general notification systems built in Domoticz.

Cheers!

**Comment by gizmocuz** [comment]
Philio is known for bad hardware (sorry philio), i am now at the moment also having (AGAIN) issues with my PAN04, my outside lights won't go out anymore !
When i unpower the node, and put the power back on, it is fixed... but for how long ?
This is the second time already in a half year... this is really annoying and has nothing todo with domoticz.
(I verified this)

I should have bought from a better manufacturer (like aeotec/fibaro), but i was doing it cheap... and cheap means problems... learned that lesson now

**Comment by RicardP** [comment]
Well I never had any problems with those PAN04 switches except for the situation I explained, when the Razberry controller put it on "Dead Node" list and nothing from node side could remove it from that status. So in my opinion as my rebooting of the Raspberry with Razberry Controller fixed it it could not be the node that caused this "non resolvable communication". 

However I fully share your opinion in cheap is generally bad, but even my Fibaro Smoke sensors at about 70 EUR/pcs(!) are having problems and till now I had to claim replacements on 3pcs out of 5... 

Ps. now I am testing latest beta with Error log e-mailing, no errors in my log though... ;-)

**Comment by gizmocuz** [comment]
To simulate an error
First, enable this option in the settings page (it is in the email setup "at the moment")
Next, add a new hardware, for example a MySensors TCP gateway, let it connect to 127.0.0.1 port 1234 (an not in use port)
You should receive errors

**Comment by RicardP** [comment]
Yes, I tested about the same procedure to initiate an error, and it worked as expected :)

**Comment by flopp999** [comment]
Nice feature, good work

**Comment by quack3d** [comment]
Should it be possible to use push notifications on errors instead of e-mail? If not, could it be? :)

**Comment by foadem** [comment]
@gizmocuz
I have an Easy ESP relay. When the device is dead for example the power is down. And I try to turn it on from domoticz switches in android app. I would like to get notified that the http command is not sent. I can see the problem in log messages which is saying:
**Error: Error opening url: http://192.168.1.121/control?cmd=gpio,2,1**
but I prefer to get notified immediately in domoticz android app. What is the solution?
